src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,218 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Diagnostics;
+using System.Text.Json.Serialization;
 
 namespace System.Text.Json
 {
     public static partial class JsonSerializer
     {
-        private static void HandleMetadataPropertyValue(ref Utf8JsonReader reader, ref ReadStack state)
+        internal static bool ResolveMetadata<T>(
+            this JsonConverter converter,
+            ref Utf8JsonReader reader,
+            ref ReadStack state,
+            out T value)
         {
-            Debug.Assert(state.Current.JsonClassInfo!.Options.ReferenceHandling.ShouldReadPreservedReferences());
+            if (state.Current.ObjectState < StackFrameObjectState.MetadataPropertyName)
+            {
+                // Read the first metadata property name.
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(converter.TypeToConvert);
+                    }
+                    else
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                    }
+                }
+
+                ReadOnlySpan<byte> propertyName = GetSpan(ref reader);
+                MetadataPropertyName metadata = GetMetadataPropertyName(propertyName);
+                state.Current.MetadataPropertyName = metadata;
+                if (metadata == MetadataPropertyName.Ref)
+                {
+                    if (!converter.CanHaveMetadata)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataInvalidPropertyWithLeadingDollarSign(propertyName, ref state, reader);
+                    }
+
+                    state.Current.ObjectState = StackFrameObjectState.MetadataRefProperty;
+                }
+                else if (metadata == MetadataPropertyName.Id)
+                {
+                    if (!converter.CanHaveMetadata)
+                    {
+                        if ((converter.ClassType & (ClassType.Dictionary | ClassType.Enumerable)) != 0)
+                        {
+                            ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(converter.TypeToConvert);
+                        }
+                        else
+                        {
+                            ThrowHelper.ThrowJsonException_MetadataInvalidReferenceToValueType(state.Current.JsonClassInfo.Type);
```

Comment:
Thanks. Changed so that `$id` are allowed for anything except immutable collections and `$ref` is allowed to anything except structs.

Yes we should add tests for structs to ensure they can have `$id`; I'll assume you'll do this after this PR based on our conversation.